src/Uno.UI/UI/Xaml/Controls/ContentDialog/DeferralManager.cs

```diff
@@ -0,0 +1,45 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace Windows.UI.Xaml.Controls
+{
+	/// <summary>
+	/// Handles completion of deferrals. The deferred action is completed when all deferral objects that were taken have called Complete().
+	/// </summary>
+	/// <typeparam name="T"></typeparam>
+	internal class DeferralManager<T> where T : IDeferral, new()
+	{
+		private readonly Action _completeHandler;
+		private int _deferralsCount;
+
+		public DeferralManager(Action completeHandler)
+		{
+			this._completeHandler = completeHandler;
+		}
+
+		public T GetDeferral()
+		{
+			_deferralsCount++;
+			var isCompleted = false;
+			return new T { DeferralAction = OnDeferralCompleted };
+
+			void OnDeferralCompleted()
+			{
+				if (isCompleted)
+				{
+					throw new InvalidOperationException("Deferral already completed.");
+				}
+
+				isCompleted = true;
+				_deferralsCount--;
+				if (_deferralsCount <= 0)
+				{
+					_completeHandler();
```

Comment:
I would assume that its legal to complete a deferral on a background thread ... so we should make sure to go back on UI thead for this

Reply:
Most of those APIs are UI Thread only, and we don't validate thread affinity yet.

Reply:
Yeah @davidjohnoliver checked it and it appears that we get an exception when completing from a BG thread ... that's sad.

Reply:
Yep. I've added a check.